### ISSUE TITLE
Bring this into VSCode proper

### ISSUE THREAD

**Issue by bpasero** [lehni/vscode-titlebar-less-macos#34]
Cross posting here from: https://github.com/microsoft/vscode/issues/12377#issuecomment-507240068

Would be great :+1:

**Comment by knopp** [comment]
Okay, given that some kind of inline titlebar might actually get into VSCode, here's my 2 cents. I recently published [extension](https://marketplace.visualstudio.com/items?itemName=iocave.customize-ui), that amongst other things also implements titlebar less on OS X (inspired by this extension, obviously :) 

There are some differences though. @lehni's extension makes activity bar very wide, which might not be the behavior you'd want in stock VSCode. It's a huge chunk of screen estate gone. My extension tucks the activity bar below the traffic lights, and if desired, let you specify background color for the titlebar (traffict lights + sidebar title). I.e. 

<img width="358" alt="Screen Shot 2019-07-07 at 20 01 58" src="https://user-images.githubusercontent.com/96958/60772142-4e6ff280-a0f2-11e9-829f-53661ee7270b.png">

<img width="360" alt="Screen Shot 2019-07-07 at 20 07 47" src="https://user-images.githubusercontent.com/96958/60772178-ee2d8080-a0f2-11e9-8054-7ae1c8004889.png">

**Comment by fabiospampinato** [comment]
> which might not be the behavior you'd want in stock VSCode.

That's exactly what I want as a user, your proposed alternative IMHO looks pretty confusing.

**Comment by knopp** [comment]
> > which might not be the behavior you'd want in stock VSCode.
> 
> That's exactly what I want as a user, your proposed alternative IMHO looks pretty confusing.

I agree it's not ideal, but neither is resizing activity bar from (already quite large) 50px to 78px. That's a lot less space for actual code.

Tbh, I'm not a fan of vertical activity bar in general, that's why I added setting to move it to the bottom, but that's a separate issue.
<img width="776" alt="Screen Shot 2019-07-07 at 20 16 41" src="https://user-images.githubusercontent.com/96958/60772277-3dc07c00-a0f4-11e9-9aa8-815493fe4179.png">

**Comment by lehni** [comment]
@bpasero I'm finally finding some time to answer this:

> @lehni I tried your extension today and must say that I am impressed by how you deal with the corner cases (activity bar hidden, side bar hidden, side bar to the right). Very well done 👍
>
> That said, would you be willing to move your changes into a PR to merge this into VSCode as core experience? The way I could see this working is by providing another window.titleBarStyle option inline. > This setting would only be available on macOS, at least until we figured out a solution for Windows/Linux.

🔗 https://github.com/microsoft/vscode/issues/12377#issuecomment-507240068

Thank you for the kind words! I'd be happy to work on integrating this into VSCode, but I'd probably need some help doing it the right away. Since we're both based in Zürich, we could even meet and discuss this over lunch or beers :)

But it sounds like VSCode needs to adopt Electron 6 first, right? I can keep maintaining the extension until then...

**Comment by bpasero** [comment]
> Thank you for the kind words! I'd be happy to work on integrating this into VSCode, but I'd probably need some help doing it the right away. 

You already did all the changes into a patch, right? We can start from that and then see if something can be done in any better way.

> Since we're both based in Zürich, we could even meet and discuss this over lunch or beers :)

Yes! though I will be out on vacation soon, but we can connect when I am back.

> But it sounds like VSCode needs to adopt Electron 6 first, right? I can keep maintaining the extension until then...

We have a branch going with Electron 6 (`electron-6.0.x`) to test things. Are you referring to a particular issue that is fixed in Electron 6 for this to work?

**Comment by lehni** [comment]
> You already did all the changes into a patch, right? We can start from that and then see if something can be done in any better way.

Yes, for that I actually worked with the VSCode source to find the right places and then condensed it into a patch. So I have some understanding of the code base. Some things are rather hacked together though :)

> Yes! though I will be out on vacation soon, but we can connect when I am back.

Happy holidays! Will you get in touch when you're back?

> We have a branch going with Electron 6 (`electron-6.0.x`) to test things. Are you referring to a particular issue that is fixed in Electron 6 for this to work?

I'm thinking of #24 that I thought was resolved in that version? Maybe I'm wrong about this.

**Comment by bpasero** [comment]
I see, yeah we can start on the PR and then see how it works out. At the same time we are working on updating to Electron 6.

I will be back beginning of August and can ping here again then.

**Comment by lehni** [comment]
@knopp it may be that the people who are happy with the status quo are also Slack users? The look was inspired by it:

![image](https://user-images.githubusercontent.com/213275/60974211-fc191680-a329-11e9-9d49-903ace1d2552.png)

But taking further cues from Slack, we could go with the traffic lights that have less padding around them, and then also make the activity bar less wide to align with it again. This would mean that the lights wouldn't align as nicely with the labels in the panes and tabs as they currently doo, see:

![image](https://user-images.githubusercontent.com/213275/60974431-5ade9000-a32a-11e9-9bec-234dfc9e1ff4.png)

**Comment by lehni** [comment]
@knopp come to think of it, this could easily be hidden behind a config switch. I'll look into this soon.

**Comment by knopp** [comment]
@lehni, I'm not a slack user so I can't say, but in my opinion horizontal screen estate for IDEs is extremely valuable and I'm not very fond of having almost 30 pixels of additional empty space just to accomodate traffic lights. 

However, it's quite possible I'm in the minority and most people are fine with sacrificing this screen estate for better esthetics :) Either way, with activity bar at the bottom this is a non issue for me.

All this said, if you go ahead with the patch, there are two issues with your extension that you might want to address:

1. Double click at any draggable area should invoke `windowService.onWindowTitleDoubleClick()` to maximize/restore window, 

2. You might also want to address problem with tabs preventing dragging through sidebar title. To reproduce, just create lot of tabs (so that they scroll horizontally), make sure that last tab is selected and try dragging through sidebar title. It will not work, because tabs have `-webkit-app-region:no-drag` set. Apparently, even if element is scrolled away and clipped, this still prevents dragging. In my extension I solved this by creating empty DOM element with `webkit-app-region:drag` that fills the space after last tab, and only in the fist row of tabs. This way, it is not necessary to have no-drag on individual tabs and thus they don't interfere with sidebar title.

You can see the code [here](https://github.com/iocave/customize-ui/blob/master/modules/title-bar.js#L174).

**Comment by fabiospampinato** [comment]
> but in my opinion horizontal screen estate for IDEs is extremely valuable and I'm not very fond of having almost 30 pixels of additional empty space just to accomodate traffic lights.

I think so too, but I also like nice UIs, the way this works for me is that I have a shortcut for toggling the activity bar, unless I actually need it (if I forget what the shortcut for switching between them is) I simply show it temporarily.

**Comment by lehni** [comment]
Personally I write code that never goes over 80 chars per line, so I can use diffing and such easily when working with git. Thus the wide activity bar never bothered me in VSCode.

I'll give the more narrow traffic lights a go now, but I don't think it's worth going further than that...

**Comment by lehni** [comment]
@knopp regarding the other things you mention, yes I'll keep them in mind. It'll be much easier to address this when the code is properly integrated in VSCode, not delivered as an increasingly complex patch.

**Comment by lehni** [comment]
Here a quick sketch of how it would look with the thiner traffic lights. I'm torn about that look, feels like it doesn't have enough room to breathe, but maybe I'm just too used to the status quo now.

![image](https://user-images.githubusercontent.com/213275/61056367-837e8c80-a3f3-11e9-90b0-e78ecfee87fe.png)

**Comment by fabiospampinato** [comment]
> but maybe I'm just too used to the status quo now.

Nah, those traffic lights just look ugly, they use a very different padding compared to other elements of the app (see "Explorer" in the sidebar or the editors' tabs), and as a consequence they stop behind aligned with anything on the x axis.

**Comment by knopp** [comment]
@fabiospampinato, out of curiosity, does this also look confusing to you?

 
<img width="820" alt="Screen Shot 2019-07-25 at 11 35 16" src="https://user-images.githubusercontent.com/96958/61863707-6a380e80-aed0-11e9-894b-2c6989a7f52e.png">

**Comment by fabiospampinato** [comment]
@knopp It still looks off to me, the window's control buttons aren't aligned with anything on the y axis, the border that separates sidebar and activity bar is kind of truncated (doesn't go all the way to the top like the border that separates the editors and the sidebar).

**Comment by knopp** [comment]
@fabiospampinato 
> window's control buttons aren't aligned with anything on the y axis,

They're center aligned by the sidebar title and tabs, so I'm not sure what you mean. Did you by any chance meant x-axis?

**Comment by fabiospampinato** [comment]
> Did you by any chance meant x-axis?

Yes, sorry 😅

Another thing that looks off to me is that the "EXPLORER" header in the sidebar isn't aligned with the elements in the sidebar. I think this is inevitable when the activity bar is hidden, but when it's visible I'd prefer if things aligned more.

**Comment by lehni** [comment]
@bpasero ~~since things appear to be breaking in a bigger way again soon (see #36)~~, is it maybe now time to start working on the real integration instead of patching it again for all the underlying changes?

**Update**: #36 seems fine after another restart.

**Comment by bpasero** [comment]
Yeah, I am a bit flooded by work right now after returning, but feel free to fire up a PR, though I may not have time to jump on it just now 👍

**Comment by lehni** [comment]
I have deprecated this extension in favour of https://marketplace.visualstudio.com/items?itemName=iocave.customize-ui by @knopp now. Feel free to carry on the discussion of integrating its functionality there!

**Comment by lehni** [comment]
@bpasero since [customize-ui is now broken](https://github.com/iocave/customize-ui/issues/156), would you still be interested bringing the changes of my extension into core? I am still based in Zurich and would be happy to work on this together. We could meet for a coffee and discuss how to best go about it! let's make this a reality. VSCode deserves a modern look!

**Comment by lehni** [comment]
@bpasero APC is now carrying the torch, but I still think this should become the default style on macOS:

https://github.com/drcika/apc-extension/issues/52

**Comment by lehni** [comment]
…and on and on it goes: APC is now also broken, but the new Custom UI Style extension is now carrying the torch. The settings to achieve this look became a bit more complex, as it's all done all in CSS only now:

https://github.com/subframe7536/vscode-custom-ui-style/issues/16
